src/mobile-app/components/app.tsx

```diff
@@ -1,21 +1,90 @@
-import React from "react";
+import React, { useEffect } from "react";
 import { useState } from "react";
 import { ExperimentPicker } from "./experiment-picker";
 import { ExperimentWrapper } from "./experiment-wrapper";
-import { IExperiment } from "../../shared/experiment-types";
+import { IExperiment, IExperimentData } from "../../shared/experiment-types";
+import { LocalDataStorage } from "../../shared/utils/local-data-storage";
+import { RunPicker } from "./run-picker";
+import { IRun } from "./types";
+import { useExperiments } from "../hooks/use-experiments";
 
 import css from "./app.module.scss";
 
-export const AppComponent: React.FC<{}> = () => {
+const runsStorage = new LocalDataStorage<IRun>("runs-");
+const availableRunKeysStorage = new LocalDataStorage<string[]>("available-runs");
 
-  const [experiment, setExperiment] = useState<IExperiment|undefined>();
+export const AppComponent: React.FC = () => {
+  const { experiments, upgradeApp } = useExperiments();
```

Comment:
I don't mind that the experiments array was "lifted" out of the component.  I did put it there for a couple of reasons:

1) to keep the data at the closest component that needed it (here you use it but then only pass it to the single component - it isn't shared by any other ones) 

2) so that the json was reloaded each time the picker was shown (thinking again about this it doesn't make operational sense as the students won't be switching between experiments very often to trigger this reload).

Reply:
I get it. If we didn't have hooks, I'd definitely prefer to keep as much of the logic as we can in the top-level component, as then it's in one place. When I lifted it out of this component, I did it for that reason. But I did it kinda forgetting that hooks keep the logic out of the component already. They add some state quite often, but most of the complexity is under the hood. Also, the logic is easier to find as they're kept together etc. On the other hand, the whole flow of the app and various state handling will be spread across many components. But we can avoid unnecessary props passing (which is annoying long term) and it's more likely we can avoid the state handling library in the end. So, I have no strong opinion here. Probably we should just pick one approach and stick to it. I think your approach is fine because of the hook advantages. I'll move the hook back here.